File: enqueue.go
package main

import (
	"bytes"
	"context"
	"encoding/json"
	"log/slog"
	"net/http"
	"os"
	"strings"
	"time"
)

// debitEnqueuer forwards failed Talos debits to the egent-jobs trigger
// endpoint (POST {EGENT_JOBS_URL}/trigger/debit-retry) so they get durable
// retry instead of becoming permanent revenue leaks. It is fire-and-forget:
// metering's OTLP export hot path must never block or fail because the retry
// queue is unreachable.
type debitEnqueuer struct {
	url     string
	hc      *http.Client
	log     *slog.Logger
	enabled bool
}

func newDebitEnqueuer(log *slog.Logger) *debitEnqueuer {
	base := strings.TrimRight(os.Getenv("EGENT_JOBS_URL"), "/")
	if base == "" {
		log.Info("debit retry enqueuer disabled (set EGENT_JOBS_URL to enable)")
		return &debitEnqueuer{log: log}
	}
	return &debitEnqueuer{
		url:     base + "/trigger/debit-retry",
		hc:      &http.Client{Timeout: 3 * time.Second},
		log:     log,
		enabled: true,
	}
}

// enqueueDebit hands a failed debit to egent-jobs for durable retry. The
// ingest payload already uses the camelCase JSON shape egent-jobs' debitretry
// Args expects (actorId/requestId/...). Best-effort: errors are logged,
// never propagated.
func (e *debitEnqueuer) enqueueDebit(req ingestRequest) {
	if !e.enabled {
		return
	}
	go func() {
		ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
		defer cancel()
		body, _ := json.Marshal(req)
		httpReq, err := http.NewRequestWithContext(ctx, http.MethodPost, e.url, bytes.NewReader(body))
		if err != nil {
			e.log.Warn("debit enqueue: build request", "requestId", req.RequestID, "error", err)
			return
		}
		httpReq.Header.Set("Content-Type", "application/json")
		resp, err := e.hc.Do(httpReq)
		if err != nil {
			e.log.Warn("debit enqueue: post failed (egent-jobs unreachable?)", "requestId", req.RequestID, "error", err)
			return
		}
		defer resp.Body.Close()
		if resp.StatusCode != http.StatusAccepted {
			e.log.Warn("debit enqueue: non-202 from egent-jobs", "status", resp.StatusCode, "requestId", req.RequestID)
			return
		}
		e.log.Info("debit enqueued for durable retry", "requestId", req.RequestID, "actorId", req.ActorID)
	}()
}

File: src/enqueue.rs
use crate::talos::IngestRequest;
use reqwest::Client;
use std::env;
use std::time::Duration;
use tokio::sync::mpsc;
use tracing::{info, warn};

/// DebitEnqueuer forwards failed Talos debits to the egent-jobs trigger
/// endpoint for durable retry. It is fire-and-forget.
#[derive(Clone)]
#[allow(dead_code)]
pub struct DebitEnqueuer {
    url: String,
    http: Client,
    enabled: bool,
    tx: mpsc::UnboundedSender<IngestRequest>,
}

impl DebitEnqueuer {
    pub fn new() -> Self {
        let base = env::var("EGENT_JOBS_URL").unwrap_or_default();
        let base = base.trim_end_matches('/').to_string();

        let (tx, rx) = mpsc::unbounded_channel::<IngestRequest>();

        if base.is_empty() {
            info!("debit retry enqueuer disabled (set EGENT_JOBS_URL to enable)");
            return Self {
                url: String::new(),
                http: Client::new(),
                enabled: false,
                tx,
            };
        }

        let url = format!("{base}/trigger/debit-retry");
        let http = Client::builder()
            .timeout(Duration::from_secs(3))
            .build()
            .unwrap();

        let enqueuer = Self {
            url: url.clone(),
            http: http.clone(),
            enabled: true,
            tx,
        };

        // Spawn background worker
        tokio::spawn(async move {
            Self::run_worker(url, http, rx).await;
        });

        enqueuer
    }

    async fn run_worker(url: String, http: Client, mut rx: mpsc::UnboundedReceiver<IngestRequest>) {
        while let Some(req) = rx.recv().await {
            if let Err(e) = Self::send_debit(&url, &http, &req).await {
                warn!(
                    requestId = %req.request_id,
                    error = %e,
                    "debit enqueue failed"
                );
            }
        }
    }

    async fn send_debit(url: &str, http: &Client, req: &IngestRequest) -> Result<(), String> {
        let resp = http
            .post(url)
            .header("Content-Type", "application/json")
            .json(req)
            .send()
            .await
            .map_err(|e| format!("post failed: {e}"))?;

        if resp.status().as_u16() != 202 {
            return Err(format!("non-202 from egent-jobs: {}", resp.status()));
        }

        info!(
            requestId = %req.request_id,
            actorId = %req.actor_id,
            "debit enqueued for durable retry"
        );

        Ok(())
    }

    /// Enqueue a failed debit for retry. Fire-and-forget.
    pub fn enqueue_debit(&self, req: IngestRequest) {
        if !self.enabled {
            return;
        }
        let _ = self.tx.send(req);
    }
}
